### PR TITLE
Resampling Holdout / time period checking during prediction

### PR DESCRIPTION
**Pull request by johannes9522**

### CODE REVIEW

File: R/LearnerForecast.R [hunk 1]
```diff
@@ -1,24 +1,42 @@
...
       span = range(task$date(row_ids)[[task$date_col]])
-      self$date_span = list(begin = span[1], end = span[2])
+      self$date_span =
+        list(begin=list(time = span[1], row_id = head(row_ids,1)), end = list(time = span[2], row_id = tail(row_ids,1)))
+      self$date_frequency = time.frequency(task$date(row_ids)[[task$date_col]])
+
+      if(length(self$date_frequency)>1){
+        warning("The timestamps are not equidistant.")
+      }
+    },
```
Comment (SteveBronder) [inline R/LearnerForecast.R]: Timestamps don't necessarily need to be equidistant. Check out chapter 2 of Advances in Financial Machine Learning. There he goes over defining the periods by the amount of ticks (trades) that happen in a stock instead of by time. So, for example, you could aggregate every 1000 trades then take the timestamp as the end of that window

File: R/ResamplingForecastHoldout.R [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      super$initialize(id = "forecast.holdout", param_set = ps, man = "mlr3::mlr_resamplings_forecast.holdout")
+    },
+
+    iters = 1L
+  ),
+
+  private = list(
+    .sample = function(ids) {
+      nr = round(length(ids) * self$param_set$values$ratio)
+      ii = ids[1:nr]
```
Comment (SteveBronder) [inline R/ResamplingForecastHoldout.R]: Here you want to check that no time IDs exist in both the training and test set so there is no leakage. If you had APPL and MSFT stock prices and accidentally put AAPL at 4:00 PM into your training and MSFT at 4:00 PM into your testing there would be time leakage across the test/training sets

File: R/LearnerForecast.R [hunk 3]
```diff
@@ -1,24 +1,42 @@
 LearnerForecast = R6Class("LearnerForecast", inherit = Learner,
```
Comment (SteveBronder) [inline R/LearnerForecast.R]: No docs?

File: R/helper.R [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  }
+  time = as.POSIXlt(time)
+  months = time$year*12 + time$mon
+  diff.months = unique(diff(months))
+  if(length(diff.months) == 1){
+    attr(diff.months, "units") = "months"
+    return(diff.months)
+  }
+  return(diff.unique)
+}
```
Comment (SteveBronder) [inline R/helper.R]: Is this for a class called `frequency` or is that just using dot in the name? I'd recommend using underscores instead of periods as periods after functions or objects are also used for S3. It may also be worth having a method for nanotime and business days

File: R/ResamplingForecastHoldout.R (hunk 2 above)
Comment (johannes9522) [reply R/ResamplingForecastHoldout.R]: I don't think I understand the issue. In this example 4:00 PM would be e.g. id = 10 for both AAPL and MSFT. The train set is then 1:10 and the test set would start at id=11.

File: R/LearnerForecast.R (hunk 1 above)
Comment (johannes9522) [reply R/LearnerForecast.R]: The purpose of this warning is to inform the user that there may be something wrong. In case of such an aggregation it can just be ignored.

File: R/ResamplingForecastHoldout.R (hunk 2 above)
Comment (SteveBronder) [reply R/ResamplingForecastHoldout.R]: If the data is wide yes, but if the data is long wouldn't the IDs be 10 and 11. See the comment [here](https://github.com/mlr-org/mlr3forecasting/pull/17#issuecomment-577402058) for an example. I only glanced at the code so if your sure there's no leakage here you can mark it resolved.
